openhantek/src/docks/SpectrumDock.cpp

```diff
@@ -16,43 +17,56 @@
 #include "utils/dsoStrings.h"
 #include "utils/printutils.h"
 
-////////////////////////////////////////////////////////////////////////////////
-// class SpectrumDock
-/// \brief Initializes the spectrum view docking window.
-/// \param settings The target settings object.
-/// \param parent The parent widget.
-/// \param flags Flags for the window manager.
-SpectrumDock::SpectrumDock(DsoSettings *settings, QWidget *parent, Qt::WindowFlags flags)
-    : QDockWidget(tr("Spectrum"), parent, flags), settings(settings) {
+template<typename... Args> struct SELECT {
+    template<typename C, typename R>
+    static constexpr auto OVERLOAD_OF( R (C::*pmf)(Args...) ) -> decltype(pmf) {
+        return pmf;
+    }
+};
+
+SpectrumDock::SpectrumDock(DsoSettingsScope *scope, QWidget *parent, Qt::WindowFlags flags)
+    : QDockWidget(tr("Spectrum"), parent, flags), scope(scope) {
 
     // Initialize lists for comboboxes
     this->magnitudeSteps = { 1e0 , 2e0 , 3e0 , 6e0 , 1e1 , 2e1 , 3e1 , 6e1 , 1e2 , 2e2 , 3e2, 6e2 };
     for (const auto& magnitude: magnitudeSteps)
         this->magnitudeStrings << valueToString(magnitude, UNIT_DECIBEL, 0);
 
-    // Initialize elements
-    for (ChannelID channel = 0; channel < settings->scope.voltage.size(); ++channel) {
-        this->magnitudeComboBox.push_back(new QComboBox());
-        this->magnitudeComboBox[channel]->addItems(this->magnitudeStrings);
-
-        this->usedCheckBox.push_back(new QCheckBox(settings->scope.voltage[channel].name));
-    }
-
     this->dockLayout = new QGridLayout();
     this->dockLayout->setColumnMinimumWidth(0, 64);
     this->dockLayout->setColumnStretch(1, 1);
-    for (ChannelID channel = 0; channel < settings->scope.voltage.size(); ++channel) {
-        this->dockLayout->addWidget(this->usedCheckBox[channel], (int)channel, 0);
-        this->dockLayout->addWidget(this->magnitudeComboBox[channel], (int)channel, 1);
 
-        // Connect signals and slots
-        connect(usedCheckBox[channel], &QCheckBox::toggled, this, &SpectrumDock::usedSwitched);
-        QComboBox* cmb = magnitudeComboBox[channel];
-        connect(cmb, static_cast<void (QComboBox::*)(int)>(&QComboBox::currentIndexChanged), this, &SpectrumDock::magnitudeSelected);
+    // Initialize elements
+    for (ChannelID channel = 0; channel < scope->voltage.size(); ++channel) {
```

Comment:
Should be scope->spectrum.size() here. The majority of code below uses scope->spectrum, except usedCheckBox which should be labelled with channel name.

Reply:
You are right. I will introduce a `scope->countChannels()` and harmonize the code in a follow-up PR.